docs/design/map/map-store-operation-offloading.md

```diff
@@ -0,0 +1,247 @@
+
+### Design
+
+#### Overall View
+When MapStore interaction is needed, interaction is executed
+asynchronously and the partition thread is freed. Once
+the interaction completes, the task/operation is rescheduled on
+the original partition thread and completes. So all the other 
+logic that updates the internals remains same.
+
+#### Configuration
+##### Configuring Offload Per Map 
+`offload` field was added in `MapStoreConfig`. It's `true` by default.    
+```java
+MapStoreConfig#setOffload
+```
+```xml
+<map-store enabled="true" initial-mode="LAZY">
+    <offload>true</offload>
+    <class-name>com.hazelcast.examples.DummyStore</class-name>
+    <write-delay-seconds>60</write-delay-seconds>
+    <write-batch-size>1000</write-batch-size>
+    <write-coalescing>true</write-coalescing>
+</map-store>
+```
+```yaml
+map-store:
+  enabled: true
+  initial-mode: LAZY
+  class-name: com.hazelcast.examples.DummyStore
+  offload: true
+  write-delay-seconds: 60
+  write-batch-size: 1000
+  write-coalescing: true
+```
+##### Configuring Offloadable Executor
+`hz:map-store-offloadable` is the name of executor
+which `Step`s are offloaded on. It can be configured
+like regular Hazelcast executors via `ExecutorConfig`
+
+#### Implementation
+We divide MapStore interaction needed operations into parts which can be executed separately.
+These self executable parts are called `Step`:
+```java
+interface Step {
+
+    void runStep(State state);
+
+    Step nextStep(State state);
+
+    boolean isOffloadStep();
+}
+```
+In its essence, a step is either an instance of a `PartitionSpecificRunnable` or a `Runnable`.
+
+And we model an operation as a sequence of `Step`s.
+
+#### Example Model for IMap#get
+For instance, sequence of get operation steps are: `READ`,`LOAD`,`ON_LOAD`,`RESPONSE`,`AFTER_RUN`. 
+
+`READ` step tries to read value in partition thread and if it finds no matching
+value in memory, `LOAD` step starts value loading from MapStore in a separate offloaded thread. 
+Upon `LOAD` finishes, execution is restarted in partition thread again from the step `ON_LOAD`.
+
+If no loading from MapStore is required, all steps are
+executed in partition thread and no offloading is done.
+
+`State` is used to pass state between `Step`s.
+
+#### Improvement for Ordering of Operations
+In previous Hazelcast versions, all map operations wait end of
+blocking MapStore operation than they progress. This provides an
+order of execution between operations. Not to break this behavior,
+in this new implementation, there are offloaded operation queues,
+these queues are per map per partition queues(so a map has a separate
+queue inside a partition) and all map operations that require MapStore
+interaction waits in offloaded queue if there is an in-flight operation
+for that specific map partition. Upon finish of the in-flight
+operation, next operation is taken from the queue and starts to execute.
+
+This change is an improvement since only a specific map's
```

Comment:
For queued operations, no new object is created, i'm using same object reference while queueing.
Current operation queues can also grow infinitely, to avoid OOME, back-pressure should be enabled.

Reply:
I also think that a new backpressure mechanism that specifically applies to these per-map queued offloaded operations would be nice improvement. Our current backpressure regulator has some deadlock issues (see: https://github.com/hazelcast/hazelcast/issues/18778#issuecomment-1194095299) since it also applies blocking on the internal operations, therefore it's not applicable for many production env.

Reply:
Better to address those issues in a different PR. 
Reason is: Back-pressure mechanism must be a general implementation and must be at a different level than data-structures. The issues of it is out of scope of the work in this PR. In this PR, we only change the place of queued operations by moving them from operation-queues to offloaded queues, before they are waiting end of map-store interaction in operation queues, now they are waiting it in offloaded-operation-queues,  i expect memory consumptions are identical with or without this PR.